tests/pytest_framework/src/syslog_ng_config/syslog_ng_config.py

```diff
@@ -54,17 +54,15 @@ def write_config_content(self):
         )
         FileIO(self.__logger_factory, self.__config_path).rewrite(rendered_config)
 
-    @staticmethod
-    def __create_statement_group(group_type, statements):
-        statement_group = StatementGroup(group_type)
-        statement_group.update_group_with_statements(cast_to_list(statements))
-        return statement_group
-
-    @staticmethod
-    def __create_logpath_group(statements=None, flags=None):
+    def __create_logpath_group(self, statements=None, flags=None):
         logpath = LogPath()
         if statements:
-            logpath.update_logpath_with_groups(cast_to_list(statements))
+            for statement in cast_to_list(statements):
+                if isinstance(statement, StatementGroup) or isinstance(statement, LogPath):
+                    logpath.add_group(statement)
+                else:
+                    group = self.create_group(statement)
+                    logpath.add_group(group)
```

Comment:
From now the __create_logpath_group() does two things: 
1, it can add statements to the logpath group
2, can create a new statement group for drivers
I would separate the second logic before calling the __create_logpath_group() -> with this we can eliminate the if statement

Other thing: the incoming "statements" parameter could be a driver or a statement, it can be confusing. 
Can we choose other name for this two entities?

Reply:
Good point thanks. Agreed on the first part, I will do so.

About the naming part, unfortunately I cannot come up with a good name. How does it sound **not** choosing a name? My suggestion for signature: `def create_logpath(items, flags=None)`? I think it would be as convenient as the original one from user point of view.

Reply:
After checking the update, I have new comment on this part.
About splitting the __create_logpath_group():
- with the new version there is an almost duplicated function: __create_logpath() vs create_logpath().
- New functions I feel a little bit complex
- however I would do this part in this way, what do you think?
```
    @staticmethod
    def __create_logpath_group(statements=None, flags=None):
        logpath = LogPath()
        if statements:
            logpath.add_groups(statements)
        if flags:
            logpath.add_flags(cast_to_list(flags))
        return logpath

    def create_statement_group(self, statements):
        statement_group = StatementGroup(statements)
        self.__syslog_ng_config["statement_groups"].append(statement_group)
        return statement_group

    def __create_statement_group_if_needed(self, statements):
        for statement in statements:
            if not isinstance(statement, (LogPath, StatementGroup)):
                self.create_statement_group(statement)

    def create_logpath(self, statements=None, flags=None):
        self.__create_statement_group_if_needed(statements)
        logpath = self.__create_logpath(statements, flags)
        self.__syslog_ng_config["logpath_groups"].append(logpath)
        return logpath

    def create_inner_logpath(self, statements=None, flags=None):
        self.__create_statement_group_if_needed(statements)
        inner_logpath = self.__create_logpath(statements, flags)
        return inner_logpath
```
- about the "statement" parameter name: we can go with this for now (later we can change it if it will confusing)

Reply:
Unfortunately I do not see the goal here. `__create_logpath()` vs `create_logpath` are not a duplicated functions (except the name, but I can rename to `__create_logpath_wrapper` or `__create_logpath_with_conversion`). However using `self.__create_statement_group_if_needed` as above is real a code duplication of the same logic in `create_logpath` and `create_inner_logpath`. If I merge the duplicated logic into a single function, then the only difference I see between the two code structures is that somehow the map function is eliminated. I believe that is intended to be eliminated  by `__create_statement_group_if_needed`, but the pseudocode is incomplete, at least I miss the aggregation of the converted `StatementGroup` into a list that would be returned. Which would be solved by the map function.

The wrapper function is a restriction/necessity that originates the logic duplication in `create_logpath` vs `create_inner_logpath`. We shortly discussed this in https://github.com/balabit/syslog-ng/pull/2443#discussion_r245476988. In the original proposal, they was only one function, which we split into two functions to simplify the api from user point of view. The original problem came from that `create_logpath` constructs a logpath instance, but also registers it as a side effect into the config, which must not be done in case of inner paths. This latter was addressed by splitting into two functions based in inner logpath-toplevel logpath, but did not eliminate the original root cause. In consequence, whenever logpath logic changed, that must be followed in both functions (inner and toplevel), as a result a function is needed for the common parts, and this is the lead I implemented above.

As an alternative, we might rethink the whole path creation and registration process, and go with the new_logpath()+add_logpath() duet, when new_logpath() would be used to construct both inner and toplevel logpaths. With that, I can avoid the wrapper function, and execute the conversion right in `new_logpath`.

Reply:
Thanks, for the explanation. After rethought my pseudo code it missed the return list (as you pointed out), and with that it will have a same complexity. This part is clear. I would add some function renaming:
```
__maybe_convert_to_statement_group() -> create_statement_group_if_needed(), 
```
```
__create_logpath() -> __create_logpath_with_conversion() 
```
and 

```
isinstance(item, StatementGroup) or isinstance(item, LogPath) -> isinstance(item, (StatementGroup, LogPath))
```

Reply:
I like the suggestions. I fixed these.